src/EditorFeatures/Core/Implementation/PasteTracking/PasteTrackingService.cs

```diff
@@ -0,0 +1,70 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Composition;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.Text;
+
+namespace Microsoft.CodeAnalysis.PasteTracking
+{
+    [Export(typeof(IPasteTrackingService)), Shared]
+    internal class PasteTrackingService : IPasteTrackingService
+    {
+        public bool TryGetPasteTrackingInformation(Document document, out PasteTrackingInformation trackingInformation)
+        {
+            if (TryGetTextBuffer(document, out var textBuffer) &&
+                textBuffer.Properties.ContainsProperty(typeof(PasteTrackingInformation)))
+            {
+                trackingInformation = textBuffer.Properties.GetProperty<PasteTrackingInformation>(typeof(PasteTrackingInformation));
+                return true;
+            }
+
+            trackingInformation = null;
+            return false;
+        }
+
+        public bool ClearPasteTrackingInformation(Document document)
+        {
+            if (!TryGetTextBuffer(document, out var textBuffer) ||
+                !textBuffer.Properties.ContainsProperty(typeof(PasteTrackingInformation)))
+            {
+                return false;
+            }
+
+            textBuffer.Properties.RemoveProperty(typeof(PasteTrackingInformation));
+
+            return true;
+        }
+
+        public void RegisterPastedTextSpan(Document document, TextSpan textSpan)
+        {
+            if (!TryGetTextBuffer(document, out var textBuffer))
+            {
+                return;
+            }
+
+            textBuffer.Changed += TextBuffer_Changed;
+            textBuffer.Properties.GetOrCreateSingletonProperty(() => new PasteTrackingInformation(textSpan));
+
+            return;
+
+            void TextBuffer_Changed(object sender, TextContentChangedEventArgs e)
+            {
+                textBuffer.Changed -= TextBuffer_Changed;
+                ClearPasteTrackingInformation(document);
+            }
```

Comment:
1. does this work if there are multiple pastes in a row.
2. i'm worried aobut the textbuffer being held onto even if the document is closed.

Reply:
1. `Changed` gets called prior to us registering a new pasted text span, so it gets cleaned up
2. Good thinking. I'll update to get the `TextView` and clear on `Closed`

Reply:
Seems the easiest way to handle this situation is to let closing all associated TextViews clear the tracking data from the buffer's properties for us.

Reply:
Look at GetOrCreatePerSubjectBufferProperty since i thnk it takes care of all of this for you.